# **Matrix Multiplication**

![](https://www.mathsisfun.com/algebra/images/matrix-multiply-ex1b.svg)

### You will work with a simple dummy data set that contains factory production values of electrical machinery parts for a corporation.

##### you will perform the following tasks:
- Load and study the data
- Extract production values as a matrix from the data
- Extract number of working days as a matrix from the data
- Multiply matrices

## **`Watch Video 5 : Matrix Multiplication`**

## Task 1 - Load and study the data

In [1]:
# Load "numpy" and "pandas" for manipulating numbers, vectors and data frames

import numpy as np
import pandas as pd

In [2]:
# Read in the "Factory_Production.csv" file as a Pandas Data Frame
# Note: Make sure the code and the data are in the same folder or specify the appropriate path for the data
df = pd.read_csv(r"E:/DataScience_and_DataAnalyst_Notes/Maths/Dataset/Factory_Production.csv",index_col=0)


In [3]:
# Take a brief look at the data frame using ".head()"
df.head()


Generators  Motors  Cables  Days
Factory                                  
A                 8      17     232   352
B                 6      20     203   348
C                10      16     187   337
D                 7      24     218   357
E                13      13     256   362

#### Feature Description:
This data set contains the production values of power units such as generators, motors and cables of different factories belonging to the same corporation.

- Factory = unique label assigned to a factory unit for the purpose of identification
- Generators = number of generators produced by the factory in a day
- Motors = number of motors produced by the factory in a day
- Cables = number of cables produced by the factory in a day
- Days = number of working days of the factory in a year

##### Study its features such as:
- The number of factory units
- The number of machine parts
- The ranges of production values

In [4]:
# Check the dimensions of the data frame using ".shape"

df.shape

(10, 4)

In [5]:
# View basic information about the data frame using ".info()"

df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, A to J
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   Generators  10 non-null     int64
 1   Motors      10 non-null     int64
 2   Cables      10 non-null     int64
 3   Days        10 non-null     int64
dtypes: int64(4)
memory usage: 400.0+ bytes


In [6]:
# View a statistical summary of the data frame using ".describe()"
# Note: Use ".transpose()" to make the summary easier to read
df.describe().transpose()


count   mean        std    min    25%    50%     75%    max
Generators   10.0    8.8   2.658320    5.0    7.0    8.5   10.75   13.0
Motors       10.0   16.4   4.926121    8.0   13.5   16.5   19.50   24.0
Cables       10.0  205.1  29.015130  156.0  187.0  206.5  221.75  256.0
Days         10.0  344.6  15.889899  316.0  337.5  347.0  355.75  365.0

##### Observations

- There are 10 rows and 4 columns
- The first 3 columns contain the production values of generators, motors and cables
- The fourth column contains the number of working days of the factories

## Task 2 - Extract production values as a matrix from the data

- The factory production values for each factory are contained in the "Generators", "Motors" and "Cables" features
- These production values will form a 10 x 3 matrix (considering all factories)

In [7]:
# Drop the "Days" column from the data frame using ".drop()" and save the data in a new data frame called "df_prod"
df_prod = df.drop('Days',axis=1)


In [9]:
# Take a brief look at the data frame "df_prod" using ".head()"
df_prod.head()


Generators  Motors  Cables
Factory                            
A                 8      17     232
B                 6      20     203
C                10      16     187
D                 7      24     218
E                13      13     256

In [10]:
# Convert the "df_pod" data frame into a matrix using "np.matrix()" and call the matrix "mat_prod"
mat_prod = np.matrix(df_prod)


In [11]:
# Print the matrix "mat_prod"

mat_prod

matrix([[  8,  17, 232],
        [  6,  20, 203],
        [ 10,  16, 187],
        [  7,  24, 218],
        [ 13,  13, 256],
        [  9,  11, 179],
        [  7,  15, 156],
        [  5,   8, 223],
        [ 12,  18, 187],
        [ 11,  22, 210]], dtype=int64)

In [12]:
# Check the dimensions of the "mat_prod" matrix using ".shape"
mat_prod.shape


(10, 3)

##### Observations

- The factory production values of generators, motors and cables of the 10 factories is a 10 x 3 matrix.
- The Numpy Matrix holds only values and does not have any indices

## Task 3 - Extract number of working days as a matrix from the data

- The number of working days for each factory is stored in the "Days" column of the data frame "df"
- These values will form a 10 x 1 matrix

In [13]:
# Convert the "Days" column of the data frame "df" into a matrix using "np.matrix()" and call the matrix "mat_days"
mat_days = np.matrix(df['Days'])


In [14]:
# Print the matrix "mat_days"

mat_days

matrix([[352, 348, 337, 357, 362, 365, 339, 324, 316, 346]], dtype=int64)

In [15]:
# Check the dimensions of the "mat_days" matrix using ".shape"
mat_days.shape


(1, 10)

##### Observations

- The number of working days of all the factories is in the form of a 1 x 10 matrix
- The Numpy Matrix holds only values and does not have any indices

## Task 4 - Multiply matrices

- The factory production values of generators, motors and cables of the 10 factories is a 10 x 3 matrix
- The number of working days of all the factories is in the form of a 1 x 10 matrix
- The product of these two matrices is equal to the linear combination of the production values vectors scaled by the number of working days

![](https://www.basic-mathematics.com/images/multiply-matrices.png)

In [16]:
# Multiply the "mat_prod" matrix by the "mat_days" matrix
# Note: The order of multiplication matters
# Note: Multiply the matrices using "mat_days * mat_prod"
mat_days * mat_prod

matrix([[ 30355,  56602, 708048]], dtype=int64)

In [19]:
# Create a Pandas Series called "total" with index as "df_prod.columns" and data as the product of the two matrices
# Note: You will have to convert the resultant matrix into an array using "np.array()" and then extract out only the list
total = pd.Series(index=df_prod.columns,data=np.array(mat_days * mat_prod)[0])


In [20]:
# Print the "total" series
total


Generators     30355
Motors         56602
Cables        708048
dtype: int64

##### Observations

- The corporation produces 30355 generators in a year
- The corporation produces 56602 motors in a year
- The corporation produces 708048 cables in a year

### Conclusions

- From the factory production data, we can calculate the total production of the corporation
- We can treat the production values of all the factories as a matrix
- We can treat the number of working days of factories as a matrix
- The product of the two matrices is equal to the linear combination of the factory production values scaled by their number of working days
- We can multiply the two matrices to calculate the total production

# **FEEDBACK**

https://forms.zohopublic.in/cloudyml/form/CloudyMLStatisticsFeedbackForm/formperma/WV946wnf0sDM_tOlH87RxZR9yMceKWGrtuPOyXzzCRc